relay-server/src/services/spooler/envelope_stack/sqlite.rs

```diff
@@ -0,0 +1,511 @@
+            self.buffer.push_back(envelope.0)
+        }
+
+        Ok(())
+    }
+
+    fn extract_envelope(&self, row: SqliteRow) -> Result<Box<Envelope>, SQLiteEnvelopeStackError> {
+        let envelope_row: Vec<u8> = row
+            .try_get("envelope")
+            .map_err(|_| SQLiteEnvelopeStackError::Empty)?;
+        let envelope_bytes = bytes::Bytes::from(envelope_row);
+        let mut envelope =
+            Envelope::parse_bytes(envelope_bytes).map_err(|_| SQLiteEnvelopeStackError::Empty)?;
+
+        let received_at: i64 = row
+            .try_get("received_at")
+            .map_err(|_| SQLiteEnvelopeStackError::Empty)?;
+        let start_time = StartTime::from_timestamp_millis(received_at as u64);
+
+        envelope.set_start_time(start_time.into_inner());
+
+        Ok(envelope)
+    }
+}
+
+impl EnvelopeStack for SQLiteEnvelopeStack {
+    type Error = SQLiteEnvelopeStackError;
+
+    async fn push(&mut self, envelope: Box<Envelope>) -> Result<(), Self::Error> {
+        if self.above_spool_threshold() {
+            self.spool_to_disk().await?;
+        }
+
+        self.buffer.push_front(envelope);
+
+        Ok(())
+    }
+
+    async fn peek(&mut self) -> Result<&Box<Envelope>, Self::Error> {
+        if self.below_unspool_threshold() {
+            self.load_from_disk().await?
+        }
+
+        self.buffer.front().ok_or(Self::Error::Empty)
+    }
+
+    async fn pop(&mut self) -> Result<Box<Envelope>, Self::Error> {
+        if self.below_unspool_threshold() {
+            self.load_from_disk().await?
+        }
+
+        self.buffer.pop_front().ok_or(Self::Error::Empty)
+    }
+}
+
+struct InsertEnvelope {
+    received_at: i64,
+    own_key: ProjectKey,
+    sampling_key: ProjectKey,
+    encoded_envelope: Vec<u8>,
+}
+
+fn build_insert_many_envelopes<'a>(
+    envelopes: impl Iterator<Item = InsertEnvelope>,
+) -> QueryBuilder<'a, Sqlite> {
+    let mut builder: QueryBuilder<Sqlite> =
+        QueryBuilder::new("INSERT INTO envelopes (received_at, own_key, sampling_key, envelope) ");
+
+    builder.push_values(envelopes, |mut b, envelope| {
+        b.push_bind(envelope.received_at)
+            .push_bind(envelope.own_key.to_string())
+            .push_bind(envelope.sampling_key.to_string())
+            .push_bind(envelope.encoded_envelope);
+    });
+
+    builder
+}
+
+pub fn build_delete_and_fetch_many_envelopes<'a>(
+    own_key: ProjectKey,
+    project_key: ProjectKey,
+    batch_size: i64,
+) -> Query<'a, Sqlite, SqliteArguments<'a>> {
+    sqlx::query(
+        "DELETE FROM
+            envelopes
+         WHERE id IN (SELECT id FROM envelopes WHERE own_key = ? AND sampling_key = ? 
+            ORDER BY received_at DESC LIMIT ?)
+         RETURNING
+            received_at, own_key, sampling_key, envelope",
+    )
+    .bind(own_key.to_string())
+    .bind(project_key.to_string())
+    .bind(batch_size)
+}
+
+fn received_at(envelope: &Envelope) -> i64 {
```

Comment:
Will have to find a better place for this, but I wanted to avoid having `ManagedEnvelope` inside the `EnvelopeStack`.

Reply:
I would also love to reduce its duplication, maybe it's enough to put the method in the `Envelope` instead of `ManagedEnvelope`.

Reply:
+1 for making this a method on `Envelope`.